### PR TITLE
Feature/BisectingKMeans 

### PR DESCRIPTION
**Pull request by AHReccese**
#### What does this implement/fix? Explain your changes.
BisectingKMeans model support
`sklearn.cluster._bisect_k_means._BisectingTree` Transporter 
fix a bug in tuple serialization in `GeneralDataStructureTransporter` Transporter
enhance the test case for `BayesianGaussianMixture` Model
fix shallow copy in `pymilo_clustering_test` func

### CODE REVIEW

**Comment by AHReccese** [comment]
`Codacy` failure is wrong since it hasn't understand conditional `import` so ignore it's failure.
https://app.codacy.com/gh/openscilab/pymilo/pullRequest?prid=13343698
@sadrasabouri @sepandhaghighi

File: pymilo/chains/clustering_chain.py [hunk 1]
```diff
@@ -5,18 +5,23 @@
...
 
-from ..pymilo_param import SKLEARN_CLUSTERING_TABLE
-
+from ..pymilo_param import SKLEARN_CLUSTERING_TABLE, NOT_SUPPORTED
 from ..exceptions.serialize_exception import PymiloSerializationException, SerilaizatoinErrorTypes
 from ..exceptions.deserialize_exception import PymiloDeserializationException, DeSerilaizatoinErrorTypes
 from traceback import format_exc
 
-
+bisecting_kmeans_support = True if SKLEARN_CLUSTERING_TABLE["BisectingKMeans"] != NOT_SUPPORTED else False
```
Comment (sadrasabouri) [inline pymilo/chains/clustering_chain.py:13]: `bisecting_kmeans_support = not (SKLEARN_CLUSTERING_TABLE["BisectingKMeans"] != NOT_SUPPORTED)`

File: tests/test_clusterings/test_clusterings.py [hunk 2]
```diff
@@ -1,29 +1,32 @@
...
 from gaussian_mixture.bayesian_gaussian_mixture import bayesian_gaussian_mixture
 from hierarchical_clustering.agglomerative_clustering import agglomerative_clustering
 from hierarchical_clustering.feature_agglomeration import feature_agglomeration
-from pymilo.pymilo_param import SKLEARN_CLUSTERING_TABLE, NOT_SUPPORTED
+
+bisecting_kmeans_support = True if SKLEARN_CLUSTERING_TABLE["BisectingKMeans"] != NOT_SUPPORTED else False
+if bisecting_kmeans_support:
+    from bisecting_kmeans import bisecting_kmeans
+
+hdbscan_support = True if SKLEARN_CLUSTERING_TABLE["HDBSCAN"] != NOT_SUPPORTED else False
```
Comment (sadrasabouri) [inline tests/test_clusterings/test_clusterings.py:21]: Same `not(X)` replacement.

File: pymilo/transporters/bisecting_tree_transporter.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
+# -*- coding: utf-8 -*-
+"""PyMilo BisectingTree(sklearn.cluster._bisect_k_means) object transporter."""
+
+from ..pymilo_param import SKLEARN_CLUSTERING_TABLE, NOT_SUPPORTED
+from .transporter import AbstractTransporter
+from .general_data_structure_transporter import GeneralDataStructureTransporter
+from ..utils.util import is_iterable
+
+bisecting_tree_support = True if SKLEARN_CLUSTERING_TABLE["BisectingKMeans"] != NOT_SUPPORTED else False
```
Comment (sadrasabouri) [inline pymilo/transporters/bisecting_tree_transporter.py:9]: Same `not(X)` replacement.

File: pymilo/transporters/bisecting_tree_transporter.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+def is_pymilo_serialized_bisecting_tree(psbt):
+    """
+    Check whether the given object is a bisecting tree which is serialized by pymilo.
+
+    :param psbt: the pymilo serialized bisecting tree object
+    :type data: dict
+    :return: boolean
+    """
+    return is_iterable(psbt) and "pymiloed_model_type" in psbt and psbt["pymiloed_model_type"] == "_BisectingTree"
```
Comment (sadrasabouri) [inline pymilo/transporters/bisecting_tree_transporter.py:116]: Put statements into three lines.

**Review by sadrasabouri — CHANGES_REQUESTED** [review]

File: pymilo/chains/clustering_chain.py (hunk 1 above)
Comment (AHReccese) [reply pymilo/chains/clustering_chain.py:13]: Actually it is `bisecting_kmeans_support = SKLEARN_CLUSTERING_TABLE["BisectingKMeans"] != NOT_SUPPORTED`
but I got your point and I will apply them all.

Comment (AHReccese) [reply pymilo/chains/clustering_chain.py]: applied.

File: tests/test_clusterings/test_clusterings.py (hunk 2 above)
Comment (AHReccese) [reply tests/test_clusterings/test_clusterings.py]: Actually it is just `X` but I got your point and I will apply them all.

File: pymilo/transporters/bisecting_tree_transporter.py (hunk 4 above)
Comment (AHReccese) [reply pymilo/transporters/bisecting_tree_transporter.py]: converted to:
```python
return (
    is_iterable(psbt) and 
    "pymilo_model_type" in psbt and 
    psbt["pymilo_model_type"] == "_BisectingTree"
)
```

**Review by sadrasabouri — APPROVED** [review]

File: pymilo/pymilo_param.py [hunk 5]
```diff
@@ -57,14 +57,20 @@
 from sklearn.cluster import FeatureAgglomeration
 from sklearn.cluster import DBSCAN
 
+bisecting_kmeans_support = False
+try:
+    from sklearn.cluster import BisectingKMeans
+    bisecting_kmeans_support = True
+except BaseException:
+    print("BisectingKmeans doesn't exist in this version of python.")
```
Comment (sepandhaghighi) [inline pymilo/pymilo_param.py:65]: The information provided in this message is not precise enough.
`BisectingKmeans doesn't exist in this version of scikit-learn`

File: pymilo/transporters/bisecting_tree_transporter.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            setattr(reconstructed_bisecting_tree, item, data[item])
+        return reconstructed_bisecting_tree
+
+
+def is_pymilo_serialized_bisecting_tree(psbt):
+    """
+    Check whether the given object is a bisecting tree which is serialized by pymilo.
+
+    :param psbt: the pymilo serialized bisecting tree object
+    :type data: dict
```
Comment (sepandhaghighi) [inline pymilo/transporters/bisecting_tree_transporter.py:113]: `:type psbt`

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]
@AHReccese Please take a look at my comments 🔥

File: pymilo/pymilo_param.py (hunk 5 above)
Comment (AHReccese) [reply pymilo/pymilo_param.py]: we decided to remove this log.

**Review by sepandhaghighi — APPROVED** [review]
🚀

**Pull request merged**